flink-formats/flink-orc/src/main/java/org/apache/flink/orc/OrcFileSystemFormatFactory.java

```diff
@@ -83,14 +96,227 @@ private static Properties getOrcProperties(ReadableConfig options) {
+					return null;
+				}
+
+				PredicateLeaf.Type colType = toOrcType(((FieldReferenceExpression) callExp.getChildren().get(0)).getOutputDataType());
+				if (colType == null) {
+					// unsupported type
+					LOG.debug("Unsupported predicate [{}] cannot be pushed into OrcTableSource.", callExp);
+					return null;
+				}
+
+				String colName = getColumnName(callExp);
+
+				if (funcDef == BuiltInFunctionDefinitions.IS_NULL) {
+					return new OrcSplitReader.IsNull(colName, colType);
+				} else if (funcDef == BuiltInFunctionDefinitions.IS_NOT_NULL) {
+					return new OrcSplitReader.Not(
+							new OrcSplitReader.IsNull(colName, colType));
+				} else {
+					OrcSplitReader.Predicate c = toOrcPredicate(callExp.getChildren().get(0));
+					if (c == null) {
+						return null;
+					} else {
+						return new OrcSplitReader.Not(c);
+					}
+				}
+			} else if (funcDef == BuiltInFunctionDefinitions.OR) {
+				if (callExp.getChildren().size() < 2) {
+					return null;
+				}
+				Expression left = callExp.getChildren().get(0);
+				Expression right = callExp.getChildren().get(1);
+
+				OrcSplitReader.Predicate c1 = toOrcPredicate(left);
+				OrcSplitReader.Predicate c2 = toOrcPredicate(right);
+				if (c1 == null || c2 == null) {
+					return null;
+				} else {
+					return new OrcSplitReader.Or(c1, c2);
+				}
+			} else {
+				if (!isBinaryValid(callExp)) {
+					// not a valid predicate
+					LOG.debug("Unsupported predicate [{}] cannot be pushed into OrcFileSystemFormatFactory.", callExp);
+					return null;
+				}
+
+				PredicateLeaf.Type litType = getLiteralType(callExp);
+				if (litType == null) {
+					// unsupported literal type
+					LOG.debug("Unsupported predicate [{}] cannot be pushed into OrcFileSystemFormatFactory.", expression);
+					return null;
+				}
+
+				boolean literalOnRight = literalOnRight(callExp);
+				String colName = getColumnName(callExp);
+
+				// fetch literal and ensure it is serializable
+				Object literalObj = getLiteral(callExp).get();
+				Serializable literal;
+				// validate that literal is serializable
+				if (literalObj instanceof Serializable) {
+					literal = (Serializable) literalObj;
+				} else {
+					LOG.warn("Encountered a non-serializable literal of type {}. " +
+									"Cannot push predicate [{}] into OrcFileSystemFormatFactory. " +
+									"This is a bug and should be reported.",
+							literalObj.getClass().getCanonicalName(), expression);
+					return null;
+				}
+
+				if (funcDef == BuiltInFunctionDefinitions.EQUALS) {
+					return new OrcSplitReader.Equals(colName, litType, literal);
+				} else if (funcDef == BuiltInFunctionDefinitions.NOT_EQUALS) {
+					return new OrcSplitReader.Not(
+							new OrcSplitReader.Equals(colName, litType, literal));
+				} else if (funcDef == BuiltInFunctionDefinitions.GREATER_THAN) {
+					if (literalOnRight) {
+						return new OrcSplitReader.Not(
+								new OrcSplitReader.LessThanEquals(colName, litType, literal));
```

Comment:
Looks like bug here? We should add more cases.

Reply:
My idea is to only support = <> > >= < <= expressions like OrcTableSource. What expressions do you think need to be supported?

Reply:
I see, you are right.